psi4/src/psi4/libmints/thc_eri.cc

```diff
@@ -0,0 +1,364 @@
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "thc_eri.h"
+#include "basisset.h"
+#include "mintshelper.h"
+#include "integral.h"
+
+#include "psi4/libqt/qt.h"
+#include "psi4/libfock/cubature.h"
+#include "psi4/lib3index/dftensor.h"
+
+#include <algorithm>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+namespace psi {
+
+THC_Computer::THC_Computer(std::shared_ptr<Molecule> molecule, std::shared_ptr<BasisSet> primary, Options& options) :
+    molecule_(molecule), primary_(primary), options_(options) {}
+
+THC_Computer::~THC_Computer() {}
+
+LS_THC_Computer::LS_THC_Computer(std::shared_ptr<Molecule> molecule, std::shared_ptr<BasisSet> primary, 
+                                    Options& options) : THC_Computer(molecule, primary, options), auxiliary_(nullptr) {}
+
+LS_THC_Computer::LS_THC_Computer(std::shared_ptr<Molecule> molecule, std::shared_ptr<BasisSet> primary, std::shared_ptr<BasisSet> auxiliary, 
+                                    Options& options) : THC_Computer(molecule, primary, options), auxiliary_(auxiliary) {}
+
+LS_THC_Computer::~LS_THC_Computer() {}
+
+void LS_THC_Computer::print_header() {
+    outfile->Printf("   --------------------------------------------\n");
+    outfile->Printf("       Least-Squares Tensor Hypercontraction   \n");
+    outfile->Printf("                   by Andy Jiang               \n");
+    outfile->Printf("              doi: 10.1063/1.4768233           \n");
+    outfile->Printf("   --------------------------------------------\n\n");
+    outfile->Printf("  Detailed LS-THC thresholds and cutoffs:\n");
+    outfile->Printf("    LS_THC_SPHERICAL_POINTS   =   %3d \n", options_.get_int("LS_THC_SPHERICAL_POINTS"));
+    outfile->Printf("    LS_THC_RADIAL_POINTS      =   %3d \n", options_.get_int("LS_THC_RADIAL_POINTS"));
+    outfile->Printf("    LS_THC_BASIS_TOLERANCE    = %6.3e \n", options_.get_double("LS_THC_BASIS_TOLERANCE"));
+    outfile->Printf("    LS_THC_WEIGHTS_TOLERANCE  = %6.3e \n", options_.get_double("LS_THC_WEIGHTS_TOLERANCE"));
+    outfile->Printf("    LS_THC_S_EPSILON          = %6.3e \n", options_.get_double("LS_THC_S_EPSILON"));
+    outfile->Printf("    Using DF?                     %3s \n", use_df_ ? "Yes" : "No");
+    outfile->Printf("\n\n");
+}
+
+/// @brief Builds the E intermediate using exact four-center two-electron integrals (Parrish et al. 2012 procedure 2)
+SharedMatrix LS_THC_Computer::build_E_exact() {
+
+    size_t nbf = primary_->nbf();
+    size_t rank = x1_->nrow();
+
+    int nthreads = 1;
+#ifdef _OPENMP
+    nthreads = Process::environment.get_n_threads();
+#endif
+
+    IntegralFactory factory(primary_);
+    std::vector<std::shared_ptr<TwoBodyAOInt>> eri_computers(nthreads);
+
+    eri_computers[0] = std::shared_ptr<TwoBodyAOInt>(factory.eri());
+#pragma omp parallel for
```

Comment:
Extremely small complaint here, but I can't imagine that this loop has so much work done in it that it justifies the time spent forking the threads for parallel execution thereof.